fiat-sql/src/main/kotlin/com/netflix/spinnaker/fiat/permissions/SqlPermissionsRepository.kt

```diff
@@ -61,59 +61,83 @@ class SqlPermissionsRepository(
-                    )
-                    .values(userId, permission.isAdmin, clock.millis())
-                    .onConflict(User.ID)
-                    .doUpdate()
-                    .set(mapOf(
-                        User.ADMIN to permission.isAdmin,
-                        User.UPDATED_AT to clock.millis(),
-                    ))
-                    .execute()
-
-                // Clear existing permissions
-                ctx
-                    .deleteFrom(Table.PERMISSION)
-                    .where(Permission.USER_ID.eq(userId))
-                    .execute()
+                putAllResources(ctx, allResourcesByType)
 
-                // Update permissions
-                val insertInto =
-                    ctx.insertInto(
-                        Table.PERMISSION,
-                        Permission.USER_ID,
-                        Permission.RESOURCE_NAME,
-                        Permission.RESOURCE_TYPE,
-                        Permission.BODY
-                    )
-
-                permission.allResources.map { r ->
-                    val body = objectMapper.writeValueAsString(r)
-                    insertInto.values(userId, r.name, r.resourceType.toString(), body)
-                }
-
-                insertInto.execute()
+                putUserPermission(ctx, permission.id, permission.isAdmin, allResourcesByType)
             }
         }
 
         return this
     }
 
     override fun putAllById(permissions: Map<String, UserPermission>?) {
-        permissions?.values?.forEach { permission -> put(permission) }
-    }
+        if (permissions == null || permissions.isEmpty()) {
+            return
+        }
+
+        val allResourcesByType = permissions.values
+            .map { it.allResources }
+            .flatten()
+            .groupBy { it.resourceType }
+            .mapValues { (_, v) -> v.toMutableSet() }
+
+        withPool(poolName) {
+
+            jooq.transactional(sqlRetryProperties.transactions) { ctx ->
+                putAllResources(ctx, allResourcesByType)
+            }
+
+            // insert/update users and permissions
+            permissions.values.forEach { p ->
+                val allResourcesByTypeForUser = p.allResources.groupBy { it.resourceType }.mapValues { (_, v) -> v.toSet() }
+
+                // transaction per-user to avoid locking too long
+                jooq.transactional(sqlRetryProperties.transactions) { ctx ->
+                    putUserPermission(ctx, p.id, p.isAdmin, allResourcesByTypeForUser)
+                }
+            }
+
+            // Tidy up deleted users and permissions
+            jooq.transactional(sqlRetryProperties.transactions) { ctx ->
+                val batch = mutableListOf<Query>()
+
+                val existingIds = ctx.select(USER.ID).from(USER).fetch(USER.ID).toSet();
+                val toDelete = existingIds.minus(permissions.keys)
 
+                if (toDelete.isNotEmpty()) {
```

Comment:
Do you know under what conditions these deletes run?

Reply:
Hm, I assumed `fiat` needed to do some tidy up in the case that service accounts were deleted. But looking at https://github.com/spinnaker/front50/blob/0887c6b2b7a25b433fff49af6f93ea79a66e173e/front50-core/src/main/java/com/netflix/spinnaker/front50/ServiceAccountsService.java#L80 it looks like `front50` will take care of triggering a sync etc.

The only thing is I think it would leave the user record around in `fiat` so that would still need to be deleted. I'll do some more digging.

Reply:
You might be right, I'm honestly just not sure.  Feels like it doesn't hurt to keep this here, just not totally sure what condition this supports.

The front50 triggered sync is conditional though, so we can't always depend on that.